RecDP/pyrecdp/core/utils.py

```diff
@@ -234,4 +236,124 @@ def update_linklist(linklist, key, value):
+        words = token_func(document)
+    else:
+        words = split_on_whitespace(document, new_line, tab)
+    return words
+
+
+def strip(document, strip_characters):
+    """
+    Way faster than document.strip(strip_characters) since strip_characters is
+    now a set instead of a str, and it contains a lot of elements (all the
+    emojis).
+
+    :param document: document to be processed
+    :param strip_characters: characters uesd for stripping document
+    :return: stripped document
+    """
+    if not document:
+        return document
+    beg_ind = 0
+    end_ind = len(document)
+    for i in range(len(document)):
+        if document[i] in strip_characters:
+            beg_ind += 1
+        else:
+            break
+    for i in range(1, len(document) + 1):
+        if document[-i] in strip_characters:
+            end_ind -= 1
+        else:
+            break
+    document_stripped = document[beg_ind:end_ind]
+    return document_stripped
+
+
+def words_augmentation(words, group_size, join_char):
+    """
+    Augment words, especially for Chinese (without a space between words) and
+    Vietnamese (with a space between syllables).
+
+    :param word: word list to be augmented
+    :param group_size: the size of word groups that need to be merged
+    :param join_char: characters to be added between word group
+    :return: word list after augment
+    """
+    augmentation = [
+        join_char.join(words[i:i + group_size])
+        for i in range(len(words) - group_size + 1)
+    ]
+    return augmentation
+
+
+def words_refinement(words,
+                     lower_case=False,
+                     strip_chars=None,
+                     use_words_aug=False,
+                     words_aug_group_sizes=[2],
+                     words_aug_join_char=''):
+    """
+    Refine split words. Non reversible since the document is split on
+    multiple characters, words are stripped of special characters and
+    characters are converted to lower case.
+
+    :param words: the word list to be augmented
+    :param lower_case: whether to convert word to lowercase
+    :param strip_chars: chars that need to be stripped in words
+    :param use_words_aug: whether to use word augmentation
+    :param words_aug_group_sizes: the size of word groups that need to
+        be merged
+    :param words_aug_join_char: characters to be added between word
+        group
+    :return: refined words or word list
+    """
+
+    if lower_case:
+        words = [word.lower() for word in words]
+    if strip_chars:
+        words = [strip(word, strip_chars) for word in words]
+        words = [word for word in words if word]
+    if use_words_aug:
+        augmentation = [
+            words_augmentation(words, group_size, words_aug_join_char)
+            for group_size in words_aug_group_sizes
+        ]
+        augmentation = [word for augm in augmentation for word in augm]
+        words = words + augmentation
+    return words
```

Comment:
Let's create a new file for all text related utils. some of them are under primitives/operations/llmutils/util.py